### PR TITLE
Closes #313 | Add dataloader for Saltik

### LINKED ISSUE #313: Create dataset loader for Saltik
Dataloader name: `saltik/saltik.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?saltik

|     Dataset| saltik |
|-------------|---|
| Description | Saltik is a dataset for benchmarking non-word error correction method accuracy in evaluating Indonesian words. It consists of 58,532 non-word errors generated from 3,000 of the most popular Indonesian words. |
| Subsets | - |
| Languages | ind |
| Tasks | Error Spelling Correction |
| License | GNU Affero General Public License v3.0 (agpl-3.0) |
| Homepage | https://github.com/ir-nlp-csui/saltik |
| HF URL | - |
| Paper URL | https://github.com/ir-nlp-csui/saltik/blob/main/README.md |

### PR DESCRIPTION
**Pull request by TysonYu**
Please name your PR after the issue it closes. You can use the following line: "Closes #313" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

**Comment by TysonYu** [comment]
Use new packages other than the template:
import jsonlines
import json

File: seacrowd/utils/constants.py [hunk 1]
```diff
@@ -129,6 +129,7 @@ class Tasks(Enum):
 
     # No seacrowd schema
     FACT_CHECKING = "FCT"
+    NON_WORD_ERROR_SPELLING_CORRECTION = "NWESC"
```
Comment (jen-santoso) [inline seacrowd/utils/constants.py:132]: About the addition in `constants.py`, update with the latest master too, there seem to be merge conflicts.

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+class NewDataset(datasets.GeneratorBasedBuilder):
+    """It consists of 58,532 non-word errors generated from 3,000 of the most popular Indonesian words."""
+
+    SOURCE_VERSION = datasets.Version(_SOURCE_VERSION)
+    SEACROWD_VERSION = datasets.Version(_SEACROWD_VERSION)
+
+
+    BUILDER_CONFIGS = [
+        SEACrowdConfig(
+            name="saltik_source",
```
Comment (jen-santoso) [inline seacrowd/sea_datasets/saltik/saltik.py:82]:
```suggestion
            name=f"{_DATASETNAME}_source",
```

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            version=SOURCE_VERSION,
+            description="saltik source schema",
```
Comment (jen-santoso) [inline seacrowd/sea_datasets/saltik/saltik.py:84]:
```suggestion
            description=f"{_DATASETNAME} source schema",
```

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            schema="source",
+            subset_id="saltik",
```
Comment (jen-santoso) [inline seacrowd/sea_datasets/saltik/saltik.py:86]:
```suggestion
            subset_id=f"{_DATASETNAME}",
```

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "saltik_source"
```
Comment (jen-santoso) [inline seacrowd/sea_datasets/saltik/saltik.py:90]:
```suggestion
    DEFAULT_CONFIG_NAME = f"{_DATASETNAME}_source"
```

**Review by jen-santoso — CHANGES_REQUESTED** [review]
Hello @TysonYu!
Thank you for the dataloader implementation.
In overall it runs perfectly, however there are several nits within the dataloader.
Also, there seem to be merge conflicts in `constants.py`, please resolve them.

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Saltik is a dataset for benchmarking non-word error correction method accuracy in evaluating Indonesian words.
+It consists of 58,532 non-word errors generated from 3,000 of the most popular Indonesian words.
+"""
+import os
```
Comment (ryanignatius) [inline seacrowd/sea_datasets/saltik/saltik.py:20]: please remove unused import

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+from typing import Dict, List, Tuple
+
+import datasets
+import jsonlines
+import json
+
+from seacrowd.utils.configs import SEACrowdConfig
+from seacrowd.utils.constants import Tasks, Licenses
+
+# TODO: Add BibTeX citation
```
Comment (ryanignatius) [inline seacrowd/sea_datasets/saltik/saltik.py:31]: please remove this todo comment

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  title     = {A Comparison Between SymSpell and a Combination of Damerau-Levenshtein Distance With the Trie Data Structure},
+  journal   = {2023 10th International Conference on Advanced Informatics: Concept, Theory and Application (ICAICTA)},
+  volume    = {},
+  year      = {2023},
+  url       = {https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=10390399&casa_token=HtJUCIGGlWYAAAAA:q8ll1RWmpHtSAq2Qp5uQAE1NJETx7tUYFZIvTO1IWoaYy4eqFETSsm9p6C7tJwLZBGq5y8zc3A&tag=1},
+  doi       = {},
+  biburl    = {https://scholar.googleusercontent.com/scholar.bib?q=info:5C2ZSuuJCDcJ:scholar.google.com/&output=citation&scisdr=ClF7-LnEEO-epgRychQ:AFWwaeYAAAAAZa50ahRiC4_sEsA-4WL6HB3SpIo&scisig=AFWwaeYAAAAAZa50aiWPCHL1NAShFbo-QupEL2c&scisf=4&ct=citation&cd=-1&hl=en},
+  bibsource = {google}
+}
+"""
```
Comment (ryanignatius) [inline seacrowd/sea_datasets/saltik/saltik.py:44]: for the citation, we can copy directly from the github: https://github.com/ir-nlp-csui/saltik?tab=readme-ov-file#references

**Comment by ryanignatius** [comment]
please also add `__init__.py` file

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+Saltik is a dataset for benchmarking non-word error correction method accuracy in evaluating Indonesian words.
+It consists of 58,532 non-word errors generated from 3,000 of the most popular Indonesian words.
+"""
+
+_HOMEPAGE = "https://github.com/ir-nlp-csui/saltik"
+
+_LANGUAGES = ["ind"]
+
+
+_LICENSE = "GNU Affero General Public License v3.0 (agpl-3.0)"
```
[the remark this answers is not in the document]
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:58]: done

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+_SUPPORTED_TASKS = [Tasks.NON_WORD_ERROR_SPELLING_CORRECTION]
+
+_SOURCE_VERSION = "1.0.0"
+
+_SEACROWD_VERSION = "1.0.0"
+
+
+class NewDataset(datasets.GeneratorBasedBuilder):
```
[the remark this answers is not in the document]
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:73]: done

File: seacrowd/sea_datasets/saltik/saltik.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Saltik is a dataset for benchmarking non-word error correction method accuracy in evaluating Indonesian words.
+It consists of 58,532 non-word errors generated from 3,000 of the most popular Indonesian words.
+"""
```
[the remark this answers is not in the document]
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:19]: done

File: seacrowd/utils/constants.py (hunk 1 above)
Comment (TysonYu) [reply seacrowd/utils/constants.py:132]: done

File: seacrowd/sea_datasets/saltik/saltik.py (hunk 2 above)
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:82]: done

File: seacrowd/sea_datasets/saltik/saltik.py (hunk 3 above)
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:84]: done

File: seacrowd/sea_datasets/saltik/saltik.py (hunk 4 above)
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:86]: done

File: seacrowd/sea_datasets/saltik/saltik.py (hunk 5 above)
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:90]: done

File: seacrowd/sea_datasets/saltik/saltik.py (hunk 6 above)
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:20]: done

File: seacrowd/sea_datasets/saltik/saltik.py (hunk 7 above)
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:31]: done

File: seacrowd/sea_datasets/saltik/saltik.py (hunk 8 above)
Comment (TysonYu) [reply seacrowd/sea_datasets/saltik/saltik.py:44]: ok changed

**Comment by TysonYu** [comment]
> please also add `__init__.py` file

**Comment by holylovenia** [comment]
A friendly review reminder for @ryanignatius and @jen-santoso.

**Comment by ryanignatius** [comment]
@holylovenia sorry for the delay,
from my side, only need this change and after that it's approved
https://github.com/SEACrowd/seacrowd-datahub/pull/387#discussion_r1521341063

**Comment by holylovenia** [comment]
> @holylovenia sorry for the delay, from my side, only need this change and after that it's approved [#387 (comment)](https://github.com/SEACrowd/seacrowd-datahub/pull/387#discussion_r1521341063)

Got it, then it should be okay since @TysonYu has addressed this request. At a glance, it seems that @TysonYu has also addressed @jen-santoso's request (CMIIW).

Could you please remove the changes to `constant.py`, @TysonYu? After that, we can merge this PR.

**Review by ryanignatius — APPROVED** [review]

**Pull request merged**